### PR TITLE
Better ways to handle 113 protocol batchpacket

### PR DESCRIPTION
**Pull request by yefengeeeeeeeeeee**

### CODE REVIEW

**Comment by lt-name** [comment]
@tachibana-kanade-angel review

**Comment by tachibana-kanade-angel** [comment]
## 🤖 AI Review

### Summary
This PR adds special handling for protocol 113 batch packets by detecting the zlib header (`0x78 0xda`) after the `0xfe` packet ID. The intent is to handle legacy 1.1 protocol packets separately. However, the implementation has critical security and correctness issues that must be addressed before merging.

### Issues Found

#### Blocking Issues
1. **Security: Bypasses encryption validation** ([RakNetPlayerSession.java:117](/root/projects/Nukkit-MOT/src/main/java/cn/nukkit/network/session/RakNetPlayerSession.java#L117))
   - The new legacy-113 branch runs **before** the normal decryption/checksum path
   - Once `decryptionCipher` is set, encrypted payload bytes can coincidentally start with `78 da`
   - This causes valid clients to be disconnected and possibly IP-blocked
   - Creates a path that processes batches **without** encryption checksum validation
   - **Fix**: Gate this branch to pre-encryption legacy login traffic, or move it after decryption

2. **Logic Error: Incorrect protocol detection** ([RakNetPlayerSession.java:117](/root/projects/Nukkit-MOT/src/main/java/cn/nukkit/network/session/RakNetPlayerSession.java#L117))
   - Treats zlib header `0x78 0xda` as proof of protocol 113 — this is **not protocol-specific**
   - RakNet protocol 8 also uses `CompressionProvider.ZLIB` with the same header
   - `Network.processBatch` uses the passed RakNet protocol to choose 1-byte vs 3-byte legacy packet headers
   - Passing hard-coded `7` can misdecode 1.2/137 login batches that use the same zlib header
   - **Fix**: Use actual protocol version detection instead of relying on compression header

#### Non-blocking Issues
1. **Code Duplication**: The branch duplicates the malformed-packet disconnect/block logic instead of reusing `processInboundBatch`

2. **Unsafe Buffer Access**: Uses absolute indexes `1` and `2` after `readUnsignedByte()`. Should use `buffer.readerIndex()` and `readerIndex + 1` for safety

3. **Code Style**: 
   - Missing spaces in `if` statement
   - Repeats `buffer.readableBytes() >= 3` check
   - Uses magic number `7` without explanation
   - Suggest creating a named helper like `isLegacy113BatchCandidate(...)`

4. **Missing Tests**: 
   - Existing tests cover `Network.processBatch(..., 7, ...)` but not the new session-level detection
   - Need regression tests for: protocol 113 login, protocol 137 login with `78 da`, encrypted-session packets that happen to start with `78 da`

### Recommendations
1. **Move the check after decryption** or add a condition to only apply during legacy login phase
2. **Use proper protocol version detection** instead of relying on compression header bytes
3. **Extract the logic into a helper method** for better readability and testability
4. **Add comprehensive tests** for edge cases, especially encrypted sessions with `78 da` prefix
5. **Reuse existing `processInboundBatch`** to avoid code duplication

### Notes
- The PR modifies only one file: `RakNetPlayerSession.java`
- The change is small (25 lines added) but has significant security implications
- `git diff --check` passed successfully
- Test execution was blocked due to sandbox network restrictions

---
🤖 AI Review

@lt-name cannot merge — has blocking issues that need to be resolved first

**Comment by yefengeeeeeeeeeee** [comment]
> ## 🤖 AI Review
> ### Summary
> This PR adds special handling for protocol 113 batch packets by detecting the zlib header (`0x78 0xda`) after the `0xfe` packet ID. The intent is to handle legacy 1.1 protocol packets separately. However, the implementation has critical security and correctness issues that must be addressed before merging.
> 
> ### Issues Found
> #### Blocking Issues
> 1. **Security: Bypasses encryption validation** ([RakNetPlayerSession.java:117](/root/projects/Nukkit-MOT/src/main/java/cn/nukkit/network/session/RakNetPlayerSession.java#L117))
>    
>    * The new legacy-113 branch runs **before** the normal decryption/checksum path
>    * Once `decryptionCipher` is set, encrypted payload bytes can coincidentally start with `78 da`
>    * This causes valid clients to be disconnected and possibly IP-blocked
>    * Creates a path that processes batches **without** encryption checksum validation
>    * **Fix**: Gate this branch to pre-encryption legacy login traffic, or move it after decryption
> 2. **Logic Error: Incorrect protocol detection** ([RakNetPlayerSession.java:117](/root/projects/Nukkit-MOT/src/main/java/cn/nukkit/network/session/RakNetPlayerSession.java#L117))
>    
>    * Treats zlib header `0x78 0xda` as proof of protocol 113 — this is **not protocol-specific**
>    * RakNet protocol 8 also uses `CompressionProvider.ZLIB` with the same header
>    * `Network.processBatch` uses the passed RakNet protocol to choose 1-byte vs 3-byte legacy packet headers
>    * Passing hard-coded `7` can misdecode 1.2/137 login batches that use the same zlib header
>    * **Fix**: Use actual protocol version detection instead of relying on compression header
> 
> #### Non-blocking Issues
> 1. **Code Duplication**: The branch duplicates the malformed-packet disconnect/block logic instead of reusing `processInboundBatch`
> 2. **Unsafe Buffer Access**: Uses absolute indexes `1` and `2` after `readUnsignedByte()`. Should use `buffer.readerIndex()` and `readerIndex + 1` for safety
> 3. **Code Style**:
>    
>    * Missing spaces in `if` statement
>    * Repeats `buffer.readableBytes() >= 3` check
>    * Uses magic number `7` without explanation
>    * Suggest creating a named helper like `isLegacy113BatchCandidate(...)`
> 4. **Missing Tests**:
>    
>    * Existing tests cover `Network.processBatch(..., 7, ...)` but not the new session-level detection
>    * Need regression tests for: protocol 113 login, protocol 137 login with `78 da`, encrypted-session packets that happen to start with `78 da`
> 
> ### Recommendations
> 1. **Move the check after decryption** or add a condition to only apply during legacy login phase
> 2. **Use proper protocol version detection** instead of relying on compression header bytes
> 3. **Extract the logic into a helper method** for better readability and testability
> 4. **Add comprehensive tests** for edge cases, especially encrypted sessions with `78 da` prefix
> 5. **Reuse existing `processInboundBatch`** to avoid code duplication
> 
> ### Notes
> * The PR modifies only one file: `RakNetPlayerSession.java`
> * The change is small (25 lines added) but has significant security implications
> * `git diff --check` passed successfully
> * Test execution was blocked due to sandbox network restrictions
> 
> 🤖 AI Review
> 
> @lt-name cannot merge — has blocking issues that need to be resolved first

Very Strange that in fact  this.channel.config().getProtocolVersion() didnt actually return the correct value ( Raknet Protocol 7) cause 113 protocol cant go into the right path .

File: src/main/java/cn/nukkit/network/session/RakNetPlayerSession.java [hunk 1]
```diff
@@ -157,6 +157,28 @@ protected void channelRead0(ChannelHandlerContext channelHandlerContext, RakMess
                 packetBuffer = new byte[buffer.readableBytes()];
             }
 
+            if ( buffer.readableBytes() >= 3 && (byte) ((buffer.getUnsignedByte(buffer.readerIndex()) & 0xff)) == (byte) 0x78  && (byte) (buffer.getUnsignedByte(buffer.readerIndex()+1) & 0xff) == (byte) 0xda && decryptionCipher == null){ //1.1的数据包，0xfe后面必定是 0x78 0xda这两个字节
```
Comment (lt-name) [inline src/main/java/cn/nukkit/network/session/RakNetPlayerSession.java:160]: This modification may break other versions that use zlib compression.

**Review by lt-name — CHANGES_REQUESTED** [review]

File: src/main/java/cn/nukkit/network/session/RakNetPlayerSession.java (hunk 1 above)
Comment (yefengeeeeeeeeeee) [reply src/main/java/cn/nukkit/network/session/RakNetPlayerSession.java:160]: I will test with 1.2 and 1.3 and see what happen

Comment (yefengeeeeeeeeeee) [reply src/main/java/cn/nukkit/network/session/RakNetPlayerSession.java:160]: > This modification may break other versions that use zlib compression.

It seems that 1.2 use the same way as the 1.1 , Anyway my problem
